### ISSUE TITLE
memory leak from sigma_clipped_stats

### ISSUE THREAD

**Issue by anticharmquark** [astropy/astropy#10553]
OS: Ubuntu 20.04
Python v 3.7
Using Spyder v4.0.1
Astropy v4.0

I have been experiencing a memory leak when I run sigma_clipped_stats on a .fits file with nan values. I';m not sure if it has to do with the warning message that comes up or with something internal, but I was able to reproduce it with the following code:
```
from astropy.io import fits 
from astropy.stats import sigma_clipped_stats

for i in range(100000):
    with fits.open("/home/david/srcinsert/templatealigned.fits") as HDUList:
        data = HDUList[0].data
        mean, med, std = sigma_clipped_stats(data, sigma = 3.0)
```
templatealigned.fits is a fits file that has nan values. I have tried it with a fits file without nan values and I do not get a memory leak. 

I also tested whether or not it was just the warning being saved by my iPython console but I couldn't produce it with other printed outputs either. I haven't been able to find this issue elsewhere.

Thank you!
David

**Comment by saimn** [comment]
Could you share the file ? In private if you prefer.

**Comment by anticharmquark** [comment]
Sure- do you just want the .py file? I will send you both the file that I
originally noticed there was a memory leak from as well as the code I used
to reproduce it. Also I’m not sure how to send privately- does relying to
the GitHub notification email post publicly?

David

**Comment by saimn** [comment]
Do you have a script to produce the test file ?
Replying by email to github's comment will post a public message here, if you want to send me a private link you can find my email in the mailmap file (https://github.com/astropy/astropy/blob/master/.mailmap).

**Comment by saimn** [comment]
I found a leak in Bottleneck, which is used by default by astropy's sigma clip functions if it is available. @anticharmquark , can you confirm that you have Bottleneck in your environment ?
The leak happens only for big-endian data (https://github.com/pydata/bottleneck/issues/351), which is the case with FITS data, and the funny thing is that Bottleneck is actually using Numpy's function for big-endian data, it seems that they have optimized functions only for little-endian.
So workarounds are either to cast your data to little-endian, or to specify `cenfunc=np.nanmedian` and `stdfunc=np.nanstd` to use Numpy directly.

**Comment by anticharmquark** [comment]
Yes I can confirm that I have bottleneck. Thank you for finding this
workaround!

David

**Comment by saimn** [comment]
No news from bottleneck... I think it would be a good idea to not use bottleneck for big-endian data. @larrybradley ?

**Comment by pllim** [comment]
FWIW, better be slower than leaking...

**Comment by larrybradley** [comment]
Sounds fine to me.  This needs to be mentioned in the docstring, perhaps with a tip on how one can convert their input data from big- to little-endian (i.e. if they need the speed).

**Comment by qwhelan** [comment]
Bottleneck maintainer here - apologies for the delay. Could you provide sample `.fits` data and the Bottleneck versions being used? The provided reproduction script is not executable due to a missing `.fits` file.

**Comment by pllim** [comment]
https://github.com/pydata/bottleneck/issues/351 has been marked as resolved upstream. Is this still an issue?

**Comment by saimn** [comment]
The fix was released in bottleneck 1.3.3 I think so should be fixed !
